### ISSUE TITLE
ejabberd stopped working

### ISSUE THREAD

**Issue by mray** [freedombox/Plinth#1013]
Ejabberd suddenly stopped running. It was enabled but not running (despite turning on+off and rebooting) When trying to upgrade the log told me:

```
....
2017-09-15 06:51:31,445 INFO Packages that will be upgraded: apt apt-utils augeas-tools base-passwd bsdutils fdisk gcc-7-base geoip-database gir1.2-freedesktop gir1.2-glib-2.0 iproute2 libapt-inst2.0 libapt-pkg5.0 libatomic1 libaugeas0 libblkid1 libbluetooth3 libcryptsetup4 libelf1 libfdisk1 libgcc1 libgirepository-1.0-1 libgl1 libglib2.0-0 libglib2.0-data libglvnd0 libglx0 libgomp1 libhogweed4 libidn11 libidn2-0 libmount1 libnettle6 libperl5.26 libpsl5 libpulse0 libselinux1 libsemanage-common libsemanage1 libsepol1 libsmartcols1 libstdc++6 libuuid1 mount perl perl-base perl-modules-5.26 python3-selinux util-linux
2017-09-15 06:51:31,448 INFO Writing dpkg log to '/var/log/unattended-upgrades/unattended-upgrades-dpkg.log'
2017-09-15 06:56:06,810 ERROR Installing the upgrades failed!
2017-09-15 06:56:06,812 ERROR error message: 'installArchives() failed'
2017-09-15 06:56:06,818 ERROR dpkg returned a error! See '/var/log/unattended-upgrades/unattended-upgrades-dpkg.log' for details
```

... so I looked into `/var/log/unattended-upgrades/unattended-upgrades-dpkg.log` and that says:

```
...
Setting up fdisk (2.29.2-5) ...
Setting up ejabberd (17.08-1) ...
Job for ejabberd.service failed because a timeout was exceeded.
See "systemctl  status ejabberd.service" and "journalctl  -xe" for details.
invoke-rc.d: initscript ejabberd, action "restart" failed.
● ejabberd.service - A distributed, fault-tolerant Jabber/XMPP server
   Loaded: loaded (/lib/systemd/system/ejabberd.service; enabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: timeout) since Sun 2017-09-17 06:43:05 UTC; 80ms ago
     Docs: https://www.process-one.net/en/ejabberd/docs/
  Process: 8395 ExecStart=/bin/sh -c /usr/sbin/ejabberdctl start && /usr/sbin/ejabberdctl started (code=killed, signal=TERM)

Sep 17 06:43:05 freedombox systemd[1]: Failed to start A distributed, fault-tolerant Jabber/XMPP server.
Sep 17 06:43:05 freedombox systemd[1]: ejabberd.service: Unit entered failed state.
Sep 17 06:43:05 freedombox systemd[1]: ejabberd.service: Failed with result 'timeout'.
dpkg: error processing package ejabberd (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 ejabberd
how-can-i-help: Unable to resolve your $HOME directory - cannot continue.
Log ended: 2017-09-17  06:43:11
```

What is broken here?

What can I do?

**Comment by mray** [comment]
I just logged in via SSH and started an apt-get upgrade and got the same error.

**Comment by JoKeyser** [comment]
I think I hit a similar, or even the same problem, also just a day ago; very annoying, as ejabberd kept trying to restart and filled the entire root partition with erl_crash_*.dump files.
In my `/var/log/ejabberd/ejabberd.log`, the relevant lines with either [error] or [critical] are these, where BOXNAME is the host name.

```
2017-09-18 19:57:40.705 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'tls_compression' will be likely ignored
2017-09-18 19:57:40.708 [error] <0.65.0>@ejabberd_config:validate_opts:1016 ignoring option 'ldap_servers' with invalid value: [localhost]
2017-09-18 19:57:40.709 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'erlang_node' will be likely ignored
2017-09-18 19:57:40.709 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'dhfile' will be likely ignored
2017-09-18 19:57:40.709 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'ciphers' will be likely ignored
2017-09-18 19:57:42.256 [critical] <0.179.0>@ejabberd_mnesia:init:74 Node name mismatch: I'm [ejabberd@localhost], the database is owned by [ejabberd@BOXNAME]
2017-09-18 19:57:42.260 [error] <0.179.0> CRASH REPORT Process <0.179.0> with 0 neighbours exited with reason: node_name_mismatch in gen_server:init_it/6 line 349
2017-09-18 19:57:42.785 [error] <0.201.0> CRASH REPORT Process <0.201.0> with 0 neighbours exited with reason: {aborted,{node_not_running,ejabberd@localhost}} in mnesia:abort/1 line 348
2017-09-18 19:57:42.786 [error] <0.189.0> Supervisor ejabberd_sup had child ejabberd_commands started with ejabberd_commands:start_link() at undefined exit with reason {aborted,{node_not_running,ejabberd@localhost}} in context start_error
2017-09-18 19:57:42.821 [error] <0.64.0> CRASH REPORT Process <0.64.0> with 0 neighbours exited with reason: {{shutdown,{failed_to_start_child,ejabberd_commands,{aborted,{node_not_running,ejabberd@localhost}}}},{ejabberd_app,start,[normal,[]]}} in application_master:init/4 line 134
```

I've been trying to add `erlang_node: "BOXNAME"` into the config file `/etc/ejabberd/ejabberd.yml`, but to no avail; it just gets ignored. I'll try and find out the upgrade's changelog, maybe that gives some clue...

**Comment by JoKeyser** [comment]
@mray Oh and I managed to successfully upgrade (and later, install) ejabberd by disabling the service first, so that dpkg doesn't choke on the failing restart: `systemctl disable ejabberd`, and then to `apt-get upgrade`. At least then the upgrade is off the table, I guess.

**Comment by JoKeyser** [comment]
Okay, for me, this worked: Add the BOX_HOST_NAME into  `/etc/ejabberd/ejabberdctl.cfg`. There is actually a default line that you can change:

`ERLANG_NODE=ejabberd@BOX_HOST_NAME`

There still remain a bunch of non-critical errors, though:

```
2017-09-18 20:38:03.084 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'tls_compression' will be likely ignored
2017-09-18 20:38:03.087 [error] <0.65.0>@ejabberd_config:validate_opts:1016 ignoring option 'ldap_servers' with invalid value: [localhost]
2017-09-18 20:38:03.087 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'dhfile' will be likely ignored
2017-09-18 20:38:03.087 [error] <0.65.0>@ejabberd_config:validate_opts:1022 unknown option 'ciphers' will be likely ignored
```

But I will only have time to look into these some days in the future, unfortunately.
Hope this helps your issue as well for the time being, @mray ?

**Comment by JoKeyser** [comment]
Maybe it actually was Debian's update, see https://tracker.debian.org/news/871174 , they introduced/changed some debconf settings..? Here what seems the relevant quote:

>   * debconf: normalize case and improved hostname validation
>   * debconf: updated templates with help from debian-l10n-en team
>   * debconf: check if admin account already exists before registering it
>   * Change ERLANG_NODE to upstream default, add debconf template and update
>     NEWS to help users through the change

Not sure which NEWS they mean, but apparently at least I missed them :).

**Comment by jvalleroy** [comment]
@JoKeyser The news file is here:
http://git.deb.at/w/pkg/ejabberd.git/blob/HEAD:/debian/NEWS

**Comment by mray** [comment]
> `ERLANG_NODE=ejabberd@BOX_HOST_NAME` in /etc/ejabberd/ejabberdctl.cfg

Didn't work for me. But then again what *exactly* is BOX_HOST_NAME?
I noticed the extreme logging, too and disabled ejabberd for now since /dev/root magically has grown to a whopping 18G already. o_0

**Comment by JoKeyser** [comment]
@mray , sorry for being cryptic; with BOX_HOST_NAME I just meant the (freedom)box's hostname, i.e. whatever you set in Plinth's configuration. According to the [NEWS](http://git.deb.at/w/pkg/ejabberd.git/blob/HEAD:/debian/NEWS), it should also work if you set this to `ERLANG_NODE=ejabberd`, without any `@hostname`.
The logging file size is pretty extreme. I simply removed all `erl_crash_*.dump` files.

**Comment by mray** [comment]
Thanks!

* removed the log files in `/var/log/ejabberd/`.
* edited `/etc/ejabberd/ejabberdctl.cfg` to `ERLANG_NODE=ejabberd`

Disk usage is back to normal and ejabberd can be started on Plinth again.
I'd like to close this issue, but it is likely that others run into the same problem – if I'm not mistaken. At least I don't know that I *triggered* this through some strange behavior.

What does it need to close this issue in your eyes?

**Comment by jvalleroy** [comment]
@mray, @JoKeyser, I think we need to update the ejabberd module to support latest ejabberd package. And provide a smooth upgrade path for users upgrading from the old package.

**Comment by JoKeyser** [comment]
@jvalleroy I agree, but don't yet (have the time to) fully understand what needs to be done. Surely we can hack our way out of this, but is there an (easy) way to avoid this in the future? I mean, even once we Plinth gets updated in Debian, we're dragging behind a long time - I guess that's only preventable if the ejabberd maintainers were aware of our needs?

**Comment by mray** [comment]
I completely trust your judgment on the matter but want to make one thing clear from my perspective:
This issue happening to almost anybody in our target audience would be nothing short of a *total disaster*. 
The growing logs would render the machine useless in no time. Nothing would prevent that, right?

Just noting.

**Comment by JoKeyser** [comment]
@mray Indeed, it's fair to call it that way. The problem, as I understand it, is that the ejabberd maintainers didn't/don't have our ("no-real-admin-present") use-case in mind. They've counted on/hoped us to read the NEWS, and edit the config file, if needed. The resulting disk-filling logging seems like a genuine bug to me... and to be fair, we're both using the Debian testing, not stable - so maybe that might not have been done there... would be nice to clarify how planned/justified their decision was - where the responsibilities likely lie in this case (@jvalleroy probably has an opinion?).

**Comment by jvalleroy** [comment]
@JoKeyser There was a new version of ejabberd uploaded yesterday, including this change:
* Do not force existing users to change their configuration

I wonder if any change is still needed with this fix?

I think the growing log is a separate problem though. I assume it could be caused by a variety of issues. We may want to limit the logging somehow, or clean up when it reaches a certain size. There is also this issue to disable logging all together:
https://github.com/freedombox/freedombox-todos/issues/5

**Comment by JoKeyser** [comment]
Note the problem with the crash dumps will be solved by the ejabberd maintainers, #1055 . I still haven't had time to check whether more action is needed here.
